pkg/storegateway/indexheader/stream_binary_reader.go

```diff
@@ -148,7 +177,113 @@ func newFileStreamBinaryReader(path string, postingOffsetsInMemSampling int, log
 		return nil, err
 	}
 
-	return r, nil
+	return r, err
+}
+
+// loadFromSparseIndexHeader load from sparse index-header on disk.
+func (r *StreamBinaryReader) loadFromSparseIndexHeader(ctx context.Context, logger log.Logger, id ulid.ULID, sparseHeadersPath string, sparseData []byte, postingOffsetsInMemSampling int) (err error) {
+	spanLog, _ := spanlogger.NewWithLogger(ctx, logger, "StreamBinaryReader.loadFromSparseIndexHeader")
+	start := time.Now()
+
+	defer func() {
+		level.Info(spanLog).Log("msg", "loaded sparse index-header from disk", "blockID", id, "path", sparseHeadersPath, "elapsed", time.Since(start))
+		spanLog.Span.Finish()
+	}()
+
+	level.Debug(logger).Log("msg", "reading from sparse index-header file", "path", sparseHeadersPath)
+
+	sparseHeaders := &indexheaderpb.Sparse{}
+
+	gzipped := bytes.NewReader(sparseData)
+	gzipReader, err := gzip.NewReader(gzipped)
+	if err != nil {
+		return fmt.Errorf("failed to create sparse index-header reader: %w", err)
+	}
+
+	sparseData, err = io.ReadAll(gzipReader)
+	if err != nil {
+		return fmt.Errorf("failed to read sparse index-header: %w", err)
+	}
+
+	if err := sparseHeaders.Unmarshal(sparseData); err != nil {
+		return fmt.Errorf("failed to decode sparse index-header file: %w", err)
+	}
+
+	r.symbols, err = streamindex.NewSymbolsFromSparseHeader(r.factory, sparseHeaders.Symbols, r.indexVersion, int(r.toc.Symbols))
+	if err != nil {
+		return fmt.Errorf("cannot load symbols from sparse index-header: %w", err)
+	}
+
+	r.postingsOffsetTable, err = streamindex.NewPostingOffsetTableFromSparseHeader(r.factory, sparseHeaders.PostingsOffsetTable, int(r.toc.PostingsOffsetTable), postingOffsetsInMemSampling)
+	if err != nil {
+		return fmt.Errorf("cannot load postings offset table from sparse index-header: %w", err)
+	}
+
+	return nil
+}
+
+// loadFromIndexHeader loads in symbols and postings offset table from the index-header.
+func (r *StreamBinaryReader) loadFromIndexHeader(ctx context.Context, logger log.Logger, cfg Config, indexLastPostingListEndBound uint64, postingOffsetsInMemSampling int) (err error) {
+	spanLog, _ := spanlogger.NewWithLogger(ctx, logger, "StreamBinaryReader.loadFromSparseIndexHeader")
+	start := time.Now()
+
+	defer func() {
+		level.Info(spanLog).Log("msg", "constructed sparse index-header", "elapsed", time.Since(start))
+		spanLog.Span.Finish()
+	}()
+
+	level.Debug(logger).Log("msg", "constructing sparse index-header")
```

Comment:
If you're using spanlog, all of the log statements in this method should use it (we get some nice events in the trace that way).